### ISSUE TITLE
[consistent-type-assertions] Option to disallow type casting everywhere except type guards

### ISSUE THREAD

**Issue by Stormtv** [typescript-eslint/typescript-eslint#3253]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
    "@typescript-eslint/consistent-type-assertions": [
      "warn",
      { "assertionStyle": "never" },
    ],
}
```

Currently we have to disable  this rule when checking inside type guards. It would be ideal to have an option to disable casting anywhere except inside type guards.
```TS
const publicationKeys = [
  'Foo',
  'Bar',
] as const

type PublicationKey = typeof publicationKeys[number]

/**
 * Checks if a given string is a publication key.
 *
 * @param key - String you wish to check it if is a key.
 * @returns Key is a PublicationKey.
 */
export function isPublicationKey(key: string): key is PublicationKey {
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions -- Needed for typeguard?
  return (publicationKeys as readonly string[]).includes(key)
}
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.2.0` |
| `@typescript-eslint/parser`        | `4.2.0` |
| `TypeScript`                       | `4.2.3` |
| `ESLint`                           | `7.23.0` |
| `node`                             | `15.12.0` |

**Comment by bradzacher** [comment]
you'd have this problem anywhere you want to use `publicationKeys` as a general purpose array - not just in type guards.

You can avoid needing to cast at all by using a better construct.
`Array.includes` is `O(n)`, `Set.has` is `O(1)`.

```ts
const publicationKeys = [
  'Foo',
  'Bar',
] as const

type PublicationKey = typeof publicationKeys[number]

const publicationKeysSet = new Set<string>(publicationKeys);

/**
 * Checks if a given string is a publication key.
 *
 * @param key - String you wish to check it if is a key.
 * @returns Key is a PublicationKey.
 */
export function isPublicationKey(key: string): key is PublicationKey {
  return publicationKeysSet.has(key)
}
```
[repl](https://www.typescriptlang.org/play?#code/MYewdgzgLgBADgVwEYBsCWwCGU3gNICmAnhDALwwDaAsAFAwwDkAYiCIwDR0OMBCmAJ050AujEylQkKHTpQicAjAAKydFhz5i5GPMUgAZvDUZsuMIRKUwCALZICAkbNpTox1Kc0XiEAMoEsBRgBADuMAFQADzQAmhgAOYAfAAUiJ4a5pYQAJQA3C4A9ABUxdzFMADCABYEwADWpGhGmDAJaABuBGAwsfEJMGikrenqZuAw9cQAdOXlMAACcIKYtpPaALQRUHGJMEQgCDChQ9W6IDDAtQ2DsM2Dw+tEs-QVCwKBCAKQMJYP4ioTJktM9yoU6AQAB5wEACWAGBBgYDeB6qDLjHxEFJTIgALl6O36OXxOP+aLG3j+AG9uDAPlAvj1Rl4sr5ItNqhJscQcnQAL4uIA)

Similarly `Object.hasOwnProperty` will be `O(1)` as well

```ts
const publicationKeys = { Foo: true, Bar: true } as const

type PublicationKey = keyof typeof publicationKeys;

/**
 * Checks if a given string is a publication key.
 *
 * @param key - String you wish to check it if is a key.
 * @returns Key is a PublicationKey.
 */
export function isPublicationKey(key: string): key is PublicationKey {
  return publicationKeys.hasOwnProperty(key)
}
```

**Comment by Stormtv** [comment]
Ahh this is a much better way of handling this, thanks @bradzacher
